server/mqtt.go

```diff
@@ -2743,6 +2738,40 @@ func (sess *mqttSession) bumpPI() uint16 {
 	return sess.ppi
 }
 
+// trackPublishRetained is invoked when a retained (QoS) message is published.
+// It need a new PI to be allocated, so we add it to the pendingPublish map,
+// with an empty value. Since cpending (not pending) is used to serialize the PI
+// mappings, we need to add this PI there as well. Make a unique key by using
+// mqttRetainedMsgsStreamName for the durable name, and PI for sseq.
+//
+// Lock held on entry
+func (sess *mqttSession) trackPublishRetained() uint16 {
+	// Make sure we initialize the tracking maps.
+	if sess.pendingPublish == nil {
+		sess.pendingPublish = make(map[uint16]*mqttPending)
+	}
+	if sess.cpending == nil {
+		sess.cpending = make(map[string]map[uint64]uint16)
+	}
+
+	pi := sess.bumpPI()
+	if pi == 0 {
+		return 0
+	}
+	sess.pendingPublish[pi] = &mqttPending{}
```

Comment:
I feel like you are missing setting the fields of `mqttPending`. I am not sure how you will be able to "clean up" when processing the PUBACK (in `untrackPublish()`).

Reply:
But you do set something is cpending:
```
    jsDur := mqttRetainedMsgsStreamName
    sseq := uint64(pi)
    sseqToPi, ok := sess.cpending[jsDur]
    if !ok {
        sseqToPi = make(map[uint64]uint16)
	sess.cpending[jsDur] = sseqToPi
    }
```

Reply:
Oh, sorry will remove. This was in preparation for persisting the PI mappings which I was doing out of `cpending` (for compactness). It does not belong in this PR.